File: python/ch01_agent/agent_ch01/core/commands.py
"""操作系统命令执行边界。"""

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True, slots=True)
class CommandResult:
    """操作系统命令执行后的统一结果。

    不直接返回 subprocess 对象，是为了让核心层不依赖具体操作系统 API。
    """
    output: str
    exit_code: int
    timed_out: bool
    truncated: bool


class CommandRunner(Protocol):
    """命令执行接口；真实实现和测试 Fake 都遵守它。"""
    def run(self, command: str, cwd: str, timeout_ms: int | None = None) -> CommandResult:
        """在受控工作目录中执行一条命令。"""

File: typescript/chapters/ch16/src/core/commands.ts
// 命令适配器与工具层之间的最小进程结果契约。
export interface CommandResult {
  readonly output: string;
  readonly exitCode: number;
  readonly timedOut: boolean;
  readonly truncated: boolean;
}

// cwd 必须由 Agent 上下文提供，命令文本不能自行扩大工作目录范围。
export interface CommandRunner {
  run(command: string, cwd: string, timeoutMs?: number): Promise<CommandResult>;
}
